network/src/block_handler.rs

```diff
@@ -0,0 +1,164 @@
+
+pub struct BlockHandler<'a>(&'a eth::Block);
+
+impl<'a> BlockHandler<'a> {
+    pub fn new(block: &'a eth::Block) -> Self {
+        Self(block)
+    }
+
+    /// Returns the timestamp for the start of the most recent day
+    pub fn timestamp(&self) -> i64 {
+        if let Some(header) = self.0.header.as_ref() {
+            if let Some(timestamp) = header.timestamp.as_ref() {
+                timestamp.seconds
+            } else {
+                panic!("Unable to find timestamp in block header!\nBlock: {:?}", self.0);
+            }
+        } else {
+            panic!("Unable to find header in block!\nBlock: {:?}", self.0);
+        }
+    }
+
+    /// Returns the timestamp for the start of the most recent day
+    pub fn days_timestamp(&self) -> i64 {
+        get_latest_day(self.timestamp())
+    }
+
+    /// Returns the timestamp for the start of the most recent hour
+    pub fn hours_timestamp(&self) -> i64 {
+        get_latest_hour(self.timestamp())
+    }
+
+    pub fn author(&self) -> Option<String> {
+        if let Some(header) = self.0.header.as_ref() {
+            Some(Address::from_slice(&header.coinbase).to_string())
+        } else {
+            None
+        }
+    }
+
+    pub fn block_size(&self) -> BigInt {
+        BigInt::from(self.0.size)
+    }
+
+    /// Get the supply for the current block
+    pub fn supply(&self) -> BigInt {
+        self.0
+            .balance_changes
+            .iter()
+            .map(|balance_change| {
+                // Only need to consider positive movements of balance change as the negative movements
+                // are just from the other side of the positive balance changes
+                get_balance_gain(balance_change)
+            })
+            .fold(BigInt::zero(), |sum, val| sum + val)
+    }
+
+    /// Gets the number of transactions seen in the block
+    pub fn transactions(&self) -> BigInt {
+        BigInt::from(self.0.transaction_traces.len() as u64)
+    }
+
+    pub fn difficulty(&self) -> BigInt {
+        if let Some(header) = self.0.header.as_ref() {
+            if let Some(difficulty) = header.difficulty.as_ref() {
+                return difficulty.deserialize();
+            }
+        }
+        BigInt::zero()
+    }
+
+    pub fn gas_used(&self) -> BigInt {
+        if let Some(header) = self.0.header.as_ref() {
+            BigInt::from(header.gas_used)
+        } else {
+            BigInt::zero()
+        }
+    }
+
+    pub fn gas_limit(&self) -> BigInt {
+        if let Some(header) = self.0.header.as_ref() {
+            BigInt::from(header.gas_limit)
+        } else {
+            BigInt::zero()
+        }
+    }
+
+    pub fn gas_price(&self) -> BigInt {
+        if let Some(header) = self.0.header.as_ref() {
+            if let Some(base_fee_per_gas) = header.base_fee_per_gas.as_ref() {
+                return base_fee_per_gas.deserialize() * self.gas_used();
+            }
+        }
+
+        BigInt::zero()
+    }
```

Comment:
What should this return? The average price paid for each unit of gas used in the block?

Base fee only applies to blocks after eip1559, and doesn't necessarily reflect the price of the gas.